rewards-distribution/ethereum/test/rewards_distribution_it.js

```diff
@@ -0,0 +1,718 @@
+            expect(secondRewardsBatchExecuted[0].args).to.have.property('distributionEvent', distributionEvent);
+            expect(secondRewardsBatchExecuted[0].args).to.have.property('batchHash', d.batchHashes[1]);
+            expect(secondRewardsBatchExecuted[0].args.batchIndex).to.be.bignumber.equal(new BN(1));
+        });
+
+        it("emits RewardDistributed events", async () => {
+            const d = await Driver.newWithContracts(owner);
+
+            await d.assignTokenToContract(d.getTotalAmount());
+
+            await d.announceDistributionEvent(distributionEvent);
+            const firstBatchResult = await d.executeBatch(distributionEvent, 0);
+            const secondBatchResult = await d.executeBatch(distributionEvent, 1);
+
+            const firstRewardDistributed = firstBatchResult.logs.filter(log => log.event === "RewardDistributed");
+            const secondRewardDistributed = secondBatchResult.logs.filter(log => log.event === "RewardDistributed");
+
+            expect(firstRewardDistributed).to.have.length(d.batches[0].length);
+            expect(secondRewardDistributed).to.have.length(d.batches[1].length);
+            firstRewardDistributed.forEach((l, i) => {
+                expect(l.args).to.have.property('distributionEvent', distributionEvent);
+                expect(l.args).to.have.property('recipient');
+                expect(l.args.recipient.toLowerCase()).to.equal(d.batches[0][i].address);
+                expect(l.args.amount).to.be.bignumber.equal(new BN(d.batches[0][i].amount));
+            });
+            secondRewardDistributed.forEach((l, i) => {
+                expect(l.args).to.have.property('distributionEvent', distributionEvent);
+                expect(l.args).to.have.property('recipient');
+                expect(l.args.recipient.toLowerCase()).to.equal(d.batches[1][i].address);
+                expect(l.args.amount).to.be.bignumber.equal(new BN(d.batches[1][i].amount));
+            });
+        });
+
+        it("supports processing batches out of order", async () => {
+            const d = await Driver.newWithContracts(owner);
+
+            await d.assignTokenToContract(d.getTotalAmount());
+
+            await d.announceDistributionEvent(distributionEvent);
+            await d.executeBatch(distributionEvent, 1);
+            await d.executeBatch(distributionEvent, 0);
+        });
+
+        it("fails if distributionEvent or was not announced, but succeeds otherwise", async () => {
+            const d = await Driver.newWithContracts(owner);
+
+            await d.assignTokenToContract(d.getBatchAmount(0));
+
+            await expectRevert(d.executeBatch(distributionEvent, 0));
+
+            await d.announceDistributionEvent(distributionEvent);
+            await d.executeBatch(distributionEvent, 0);
+        });
+
+        it("fails if batch or was not announced at exact position, but succeeds for declared batches", async () => {
```

Comment:
why do we need the "but" section?

Reply:
'but' section was meant to demonstrate success to indicate the cause for failure was only the deliberate corruptions of the batch data.
Removing it but adding assertion that the error received indicates a hash mismatch:
```
            const err1 = await expectRevert(d.executeBatch(distributionEvent, 0, d.batches[0].slice(1))); // wrong batch data
            expect(err1).to.have.property('reason', 'batch hash does not match');

            const err2 = await expectRevert(d.executeBatch(distributionEvent, 1, d.batches[0])); // wrong batchIndex
            expect(err2).to.have.property('reason', 'batch hash does not match');
```